[WPE] WPE Platform: add support for tap and drag gestures

Scony opened the pull request:
#### 23d6925f4d0973911cea79158805b5e555a6a0de
<pre>
[WPE] WPE Platform: add support for tap and drag gestures
<a href="https://bugs.webkit.org/show_bug.cgi?id=265637">https://bugs.webkit.org/show_bug.cgi?id=265637</a>

Reviewed by Carlos Garcia Campos.

This change extends the WPE platform API by introducing:
 - WPEGestureController - an interface (with default cross-WPE-platform implementation) that is used to process touch events in order to detect gestures.
This change also provides a code that:
 - Feeds WPEGestureController provided by WPE platform with events.
 - Handles events detected by WPEGestureController:
   - Tap gesture is handled as a mouse click.
   - Drag gesture is handled as a mouse wheel scroll.

* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/libwpe/WebEventFactory.h:
* Source/WebKit/Shared/wpe/NativeWebWheelEventWPE.cpp:
(WebKit::NativeWebWheelEvent::NativeWebWheelEvent):
* Source/WebKit/Shared/wpe/WebEventFactoryWPE.cpp:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::ViewPlatform):
(WKWPE::ViewPlatform::handleGesture):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp: Added.
(wpe_gesture_controller_default_init):
(wpe_gesture_controller_handle_event):
(wpe_gesture_controller_cancel):
(wpe_gesture_controller_get_gesture):
(wpe_gesture_controller_get_gesture_position):
(wpe_gesture_controller_get_gesture_delta):
(wpe_gesture_controller_is_drag_begin):
* Source/WebKit/WPEPlatform/wpe/WPEGestureController.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.cpp: Added.
(wpeHandleEvent):
(wpeCancel):
(wpeGetGesture):
(wpeGetGesturePosition):
(wpeGetGestureDelta):
(wpeIsDragBegin):
(wpe_gesture_controller_impl_class_init):
(wpe_gesture_controller_interface_init):
(wpeGestureControllerImplNew):
* Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.cpp: Added.
(WPE::GestureDetector::handleEvent):
(WPE::GestureDetector::reset):
* Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_set_gesture_controller):
(wpe_view_get_gesture_controller):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: <a href="https://commits.webkit.org/280661@main">https://commits.webkit.org/280661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c95c4d8891b3127c0a71766489a7e5ad98f05243

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57130 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7573 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44082 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7763 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46263 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5332 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59160 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34227 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31007 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6578 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52967 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62431 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1043 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7017 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53524 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1047 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49369 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53584 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/884 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8538 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32287 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33372 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33118 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -213,9 +213,57 @@ void PageClientImpl::doneWithKeyEvent(const NativeWebKeyboardEvent&, bool)
+        // If the touch event was handled, we must interrupt any gesture detection sequence ongoing so that gestures
+        // that require immediate feedback (e.g. drag) are not detected.
+        if (auto gestureController = wpe_view_get_gesture_controller(m_view.wpeView()))
+            wpe_gesture_controller_interrupt_gesture_detection(gestureController);
```
You should null-check first `m_view.wpeView()` that returns nullptr when using the old api

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -213,9 +213,57 @@ void PageClientImpl::doneWithKeyEvent(const NativeWebKeyboardEvent&, bool)
+#if ENABLE(WPE_PLATFORM)
+    if (auto gestureController = wpe_view_get_gesture_controller(m_view.wpeView())) {
+        auto& page = m_view.page();
+        // Some gestures don't require immediate feedback, so they are processed only if touch event was not handled.
```
I think we could handle all gestures the same way, like the gtk port does. Tap is handled on up, so at that point if thew down was handled by the web process, the sequence has already been cancelled.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -213,9 +213,57 @@ void PageClientImpl::doneWithKeyEvent(const NativeWebKeyboardEvent&, bool)
+            if (double x, y; wpe_gesture_controller_get_detected_gesture_position(gestureController, &x, &y)) {
+                // Mouse motion towards the point of the click.
+                {
+                    WPEEvent* simulatedEvent = wpe_event_pointer_move_new(
```
GRefPtr

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -213,9 +213,57 @@ void PageClientImpl::doneWithKeyEvent(const NativeWebKeyboardEvent&, bool)
+                // Mouse motion towards the point of the click.
+                {
+                    WPEEvent* simulatedEvent = wpe_event_pointer_move_new(
+                        WPE_EVENT_POINTER_MOVE, m_view.wpeView(), WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), x, y, 0, 0
```
Use `WPE_INPUT_SOURCE_TOUCHSCREEN`

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -206,15 +206,17 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
             case WPE_EVENT_TOUCH_DOWN:
-                // FIXME: gestures
+                if (auto gestureController = wpe_view_get_gesture_controller(webView.wpeView()))
+                    wpe_gesture_controller_detect_gesture(gestureController, event);
```
This should be done after sending the touch event to the web process. Also use `view` directly, since it's a parameter of the callback, you don't need to get it from the web view.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+
+namespace WPE {
+
+class GestureDetector final {
```
I think this could be in its own class.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+
+class GestureDetector final {
+public:
+    WPEGesture detectGesture(WPEEvent* event)
```
Instead of detect gesture and return a type, I think this could be just `handleEvent()` and just update the internal state. Then getting the gesture type would be a different getter to get its current state like getting the position.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+        switch (wpe_event_get_event_type(event)) {
+        case WPE_EVENT_TOUCH_DOWN:
+            if (double x, y; wpe_event_get_position(event, &x, &y))
+                m_touchDownPoint = Point { x, y };
```
I would just call this `m_origin` or something like that, touch is redundant I think. We should probably save the sequence id, to check in future events that they belong to the current sequence.

carlosgcampos replied:
I also think we can assume this is a TAP, and change it later if it ends up being a drag. Since we handle the gestures on UP, it's should be ok.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+            break;
+        case WPE_EVENT_TOUCH_MOVE:
+            if (!m_touchDownPoint)
+                break;
```
And he we check if there's a sequence id and it matches the event.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+                break;
+            if (double x, y; wpe_event_get_position(event, &x, &y)) {
+                if (!m_dragReferencePoint && std::hypot(x - m_touchDownPoint->x, y - m_touchDownPoint->y) > dragActivationThresholdPx)
+                    m_dragReferencePoint = m_touchDownPoint;
```
I think this could be just `m_position` as the current position of the current gesture, no matter it's drag or not.

carlosgcampos replied:
Here could also set the gesture as DRAG

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+            if (double x, y; wpe_event_get_position(event, &x, &y)) {
+                if (!m_dragReferencePoint && std::hypot(x - m_touchDownPoint->x, y - m_touchDownPoint->y) > dragActivationThresholdPx)
+                    m_dragReferencePoint = m_touchDownPoint;
+                if (m_dragReferencePoint) {
```
And here check if current gesture is DRAG

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+            break;
+        case WPE_EVENT_TOUCH_UP:
+            if (double x, y; wpe_event_get_position(event, &x, &y) && m_touchDownPoint) {
+                if (m_dragReferencePoint) {
```
Same here, check the gesture type instead of the points, which is easier to read

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+            reset();
+            break;
+        default:
+            reset();
```
Here we could `RELEASE_ASSERT_NOT_REACHED()`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+    void reset()
+    {
+        m_touchDownPoint = std::nullopt;
+        m_dragReferencePoint = std::nullopt;
```
Shouldn't we reset the delta too?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = { x - m_dragReferencePoint->x, y - m_dragReferencePoint->y };
```
Again, I would simplify this as `m_delta`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+                    m_dragReferencePoint = m_touchDownPoint;
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
```
Why do we need to duplicate this point?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+                    m_detectedGestureDelta = std::nullopt;
+                }
+            }
+            reset();
```
I think it would be simpler if we reset in the next down, so that after UP we can query the gesture controller.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+        return m_detectedGesture;
+    }
+
+    WPEGesture getDetectedGesture() const { return m_detectedGesture; }
```
We don't use get prefix for getters in WebKit, this could just be `gesture()`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+    using Delta = Point;
+
+    std::optional<Point> getDetectedGesturePoint() const { return m_detectedGesturePoint; }
+    std::optional<Delta> getDetectedGestureDelta() const { return m_detectedGestureDelta; }
```
Same here these could be `position()` and `delta()`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+
+    std::optional<Point> m_touchDownPoint;
+    std::optional<Point> m_dragReferencePoint;
+    WPEGesture m_detectedGesture = WPE_GESTURE_NONE;
```
```suggestion
    WPEGesture m_detectedGesture { WPE_GESTURE_NONE };
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+    WPE::GestureDetector detector;
+};
+
+WEBKIT_DEFINE_TYPE(WPEGestureController, wpe_gesture_controller, G_TYPE_OBJECT)
```
There's no common code in the base class for this, so I think this should be an interface instead of an abstract class. We could add `WPEGestureConstrollerImpl` (as private class) with the default implementation to be used by WPEDisplay by default.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ *
+ * Returns: a #WPEGesture
+ */
+WPEGesture wpe_gesture_controller_detect_gesture(WPEGestureController* controller, WPEEvent* event)
```
```suggestion
void wpe_gesture_controller_handle_event(WPEGestureController* controller, WPEEvent* event)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+
+    auto* controllerClass = WPE_GESTURE_CONTROLLER_GET_CLASS(controller);
+    if (controllerClass->get_detected_gesture)
+        return controllerClass->detect_gesture(controller, event);
```
This should be mandatory in the interface (like a pure virtual in C++)

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ *
+ * Interrupt ongoing gesture detection if any.
+ */
+void wpe_gesture_controller_interrupt_gesture_detection(WPEGestureController* controller)
```
```suggestion
void wpe_gesture_controller_cancel(WPEGestureController* controller)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+{
+    auto* controllerClass = WPE_GESTURE_CONTROLLER_GET_CLASS(controller);
+    if (controllerClass->get_detected_gesture)
+        controllerClass->interrupt_gesture_detection(controller);
```
Same here, actually I think all vfuncs should be mandatory in the interface.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ *
+ * Returns: a #WPEGesture
+ */
+WPEGesture wpe_gesture_controller_get_detected_gesture(WPEGestureController* controller)
```
```suggestion
WPEGesture wpe_gesture_controller_get_gesture(WPEGestureController* controller)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ * @x: (out): location to store x coordinate
+ * @y: (out): location to store y coordinate
+ *
+ * Get the position of previously detected gesture. If it doesn't have
```
```suggestion
 * Get the position of currently detected gesture. If it doesn't have
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ * Returns: %TRUE if position is returned in @x and @y,
+ *    or %FALSE if previously detected gesture doesn't have a positon
+ */
+gboolean wpe_gesture_controller_get_detected_gesture_position(WPEGestureController* controller, double* x, double* y)
```
```suggestion
gboolean wpe_gesture_controller_get_position(WPEGestureController* controller, double* x, double* y)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ * @x: (out): location to store delta on x axis
+ * @y: (out): location to store delta on y axis
+ *
+ * Get the delta of previously detected gesture such as "drag" gesture.
```
```suggestion
 * Get the delta of currently detected gesture such as "drag" gesture.
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,269 @@
+ * Returns: %TRUE if delta is returned in @x and @y,
+ *    or %FALSE if previously detected gesture doesn't have a delta
+ */
+gboolean wpe_gesture_controller_get_detected_gesture_delta(WPEGestureController* controller, double* x, double* y)
```
```suggestion
gboolean wpe_gesture_controller_get_delta(WPEGestureController* controller, double* x, double* y)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.h:
```diff
@@ -0,0 +1,86 @@
+ * @WPE_GESTURE_NONE:
+ * @WPE_GESTURE_TAP:
+ * @WPE_GESTURE_DRAG:
+ */
```
These should be documented

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -58,6 +59,7 @@ struct _WPEViewPrivate {
         guint button { 0 };
         guint32 time { 0 };
     } lastButtonPress;
+    GRefPtr<WPEGestureController> gestureController { WPE_GESTURE_CONTROLLER(g_object_new(WPE_TYPE_GESTURE_CONTROLLER, nullptr)) };
```
This should be initialized on demand, we don't want to create the controller if touch events are never going to be created. We should also use `std::optional` to know whether this is explicitly set to nullptr or not yet initialized

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -178,13 +180,25 @@ static void wpeViewDispose(GObject* object)
     objectClass->constructed = wpeViewConstructed;
     objectClass->dispose = wpeViewDispose;
+    viewClass->set_gesture_controller = wpeViewSetGestureController;
+    viewClass->get_gesture_controller = wpeViewGetGestureController;
```
We don't want these to be vfuncs, since the gesture controller doesn't depend on the platform implementation.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+/**
+ * wpe_view_set_gesture_controller:
+ * @view: a #WPEView
+ * @controller: a #WPEGestureController
```
```suggestion
 * @controller: (nullable) a #WPEGestureController, or %NULL
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+ * @view: a #WPEView
+ * @controller: a #WPEGestureController
+ *
+ * Set @controller as #WPEGestureController of @view
```
We should document that passing `NULL` means disabling gestures on the view.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+
+    auto* viewClass = WPE_VIEW_GET_CLASS(view);
+    if (viewClass->set_gesture_controller)
+        viewClass->set_gesture_controller(view, controller);
```
And this should be a normal setter, but we should take a reference, not adopt it.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+ */
+void wpe_view_set_gesture_controller(WPEView* view, WPEGestureController* controller)
+{
+    g_return_if_fail(WPE_IS_VIEW(view));
```
Add another g_return macro to check that controller is either null or `WPE_IS_GESTURE_CONTROLLER`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+        viewClass->set_gesture_controller(view, controller);
+}
+
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1014,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+    auto* viewClass = WPE_VIEW_GET_CLASS(view);
+    if (viewClass->get_gesture_controller)
+        return viewClass->get_gesture_controller(view);
+    return nullptr;
```
And here we create the default if the optional is nullopt

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.h:
```diff
@@ -45,6 +45,7 @@ typedef struct _WPEDisplay WPEDisplay;
 typedef struct _WPEEvent WPEEvent;
 typedef struct _WPEMonitor WPEMonitor;
 typedef struct _WPERectangle WPERectangle;
+typedef struct _WPEGestureController WPEGestureController;
```
We can include the header, since the gesture controller doesn't reference the view

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
```diff
@@ -40,10 +40,12 @@
+#if ENABLE(WPE_PLATFORM)
+#include <wpe/wpe-platform.h>
+#endif
+
```
We won't need this if we don't pass the WPEGesture to the function

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -224,7 +226,8 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
             case WPE_EVENT_TOUCH_MOVE:
-                // FIXME: gestures
+                if (auto gestureController = wpe_view_get_gesture_controller(webView.wpeView()))
+                    webView.handleGesture(wpe_gesture_controller_detect_gesture(gestureController, event), *gestureController);
```
I think it would be simpler if make `handleGesture()` receive only the event, call it unconditionally in the three cases and handle everything there.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +802,26 @@ WebKitWebViewAccessible* View::accessible() const
+#if ENABLE(WPE_PLATFORM)
+void View::handleGesture(WPEGesture gesture, WPEGestureController& controller)
+{
+    // Some gestures require immediate feedback, so they are detected and handled right away.
```
Here we would check if the view has a controller and return early otherwise.

carlosgcampos replied:
Then we call `wpe_gesture_controller_handle_event()` and then query the controller depending on the pahse we are (down, move, up, cancel)

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +802,26 @@ WebKitWebViewAccessible* View::accessible() const
+        if (double x, y, dx, dy; wpe_gesture_controller_get_detected_gesture_position(&controller, &x, &y) && wpe_gesture_controller_get_detected_gesture_delta(&controller, &dx, &dy)) {
+            auto* simulatedScrollEvent = wpe_event_scroll_new(
+                m_wpeView.get(), WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), dx, dy, TRUE, FALSE, x, y
+            );
```
We need to detect somehow the beginning of the drag to make sure that the generated event uses `WebWheelEvent::Phase::Began`.

carlosgcampos replied:
Use GRefPtr and change the source to touchscreen.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -213,8 +217,17 @@ void PageClientImpl::doneWithKeyEvent(const NativeWebKeyboardEvent&, bool)
+        // are not detected by engine itself.
+        if (!m_view.wpeView())
+            return;
+        if (auto gestureController = wpe_view_get_gesture_controller(m_view.wpeView()))
```
```suggestion
        if (auto* gestureController = wpe_view_get_gesture_controller(m_view.wpeView()))
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -206,29 +206,39 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
                 webView.page().handleTouchEvent(NativeWebTouchEvent(event, webView.touchPointsForEvent(event)));
 #endif
+                // if (auto gestureController = wpe_view_get_gesture_controller(view))
+                //     wpe_gesture_controller_handle_event(gestureController, event);
```
You forgot to remove this

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -206,29 +206,39 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
+                // if (auto gestureController = wpe_view_get_gesture_controller(view)) {
+                //     wpe_gesture_controller_handle_event(gestureController, event);
+                //     webView.handleGesture(wpe_gesture_controller_get_gesture(gestureController), *gestureController);
+                // }
```
And this

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -206,29 +206,39 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
+                // if (auto gestureController = wpe_view_get_gesture_controller(view)) {
+                //     wpe_gesture_controller_handle_event(gestureController, event);
+                //     webView.handleGesture(wpe_gesture_controller_get_gesture(gestureController), *gestureController);
+                // }
```
And this :-)

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +809,60 @@ WebKitWebViewAccessible* View::accessible() const
+#if ENABLE(WPE_PLATFORM)
+void View::handleGesture(WPEEvent* event)
+{
+    auto gestureController = wpe_view_get_gesture_controller(m_wpeView.get());
```
```suggestion
    auto* gestureController = wpe_view_get_gesture_controller(m_wpeView.get());
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +809,60 @@ WebKitWebViewAccessible* View::accessible() const
+            {
+                GRefPtr<WPEEvent> simulatedEvent { wpe_event_pointer_move_new(
+                    WPE_EVENT_POINTER_MOVE, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, static_cast<WPEModifiers>(0), x, y, 0, 0
+                ) };
```
```suggestion
                GRefPtr<WPEEvent> simulatedEvent = adoptGRef(wpe_event_pointer_move_new(
                    WPE_EVENT_POINTER_MOVE, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, static_cast<WPEModifiers>(0), x, y, 0, 0
                ));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +809,60 @@ WebKitWebViewAccessible* View::accessible() const
+            {
+                GRefPtr<WPEEvent> simulatedEvent { wpe_event_pointer_button_new(
+                    WPE_EVENT_POINTER_DOWN, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, WPE_MODIFIER_POINTER_BUTTON1, 1, x, y, 1
+                ) };
```
```suggestion
                GRefPtr<WPEEvent> simulatedEvent = adoptGRef(wpe_event_pointer_button_new(
                    WPE_EVENT_POINTER_DOWN, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, WPE_MODIFIER_POINTER_BUTTON1, 1, x, y, 1
                ));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +809,60 @@ WebKitWebViewAccessible* View::accessible() const
+            {
+                GRefPtr<WPEEvent> simulatedEvent { wpe_event_pointer_button_new(
+                    WPE_EVENT_POINTER_UP, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, static_cast<WPEModifiers>(0), 1, x, y, 0
+                ) };
```
```suggestion
                GRefPtr<WPEEvent> simulatedEvent = adoptGRef(wpe_event_pointer_button_new(
                    WPE_EVENT_POINTER_UP, m_wpeView.get(), WPE_INPUT_SOURCE_TOUCHSCREEN, 0, static_cast<WPEModifiers>(0), 1, x, y, 0
                ));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -799,6 +809,60 @@ WebKitWebViewAccessible* View::accessible() const
+    case WPE_GESTURE_NONE:
+        break;
+    case WPE_GESTURE_TAP:
+        if (double x, y; wpe_gesture_controller_get_gesture_position(gestureController, &x, &y)) {
```
I guess we only want to do this is the event is UP.

Scony replied:
Yes, there's check for that above.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGesture.h:
```diff
@@ -0,0 +1,42 @@
+
+    WPE_GESTURE_TAP,
+    WPE_GESTURE_DRAG,
+} WPEGesture;
```
I guess this could be in the controller header

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -1000,3 +1003,38 @@ void wpe_view_set_opaque_rectangles(WPEView* view, WPERectangle* rects, guint re
+    g_return_if_fail(WPE_IS_VIEW(view));
+    g_return_if_fail(WPE_IS_GESTURE_CONTROLLER(controller));
+
+    view->priv->gestureController = adoptGRef(controller);
```
This shouldn't take the ownership, just take a reference.

carlosgcampos commented on Source/WebKit/Shared/NativeWebWheelEvent.h:
```diff
@@ -68,6 +68,7 @@ class NativeWebWheelEvent : public WebWheelEvent {
     NativeWebWheelEvent(struct wpe_input_axis_event*, float deviceScaleFactor, WebWheelEvent::Phase, WebWheelEvent::Phase momentumPhase);
 #if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
     explicit NativeWebWheelEvent(WPEEvent*);
+    explicit NativeWebWheelEvent(WPEEvent*, WebWheelEvent::Phase);
```
```suggestion
    NativeWebWheelEvent(WPEEvent*, WebWheelEvent::Phase);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
```diff
@@ -219,8 +223,15 @@ void PageClientImpl::doneWithTouchEvent(const NativeWebTouchEvent& touchEvent, b
+        // If the touch event was handled, we must interrupt any gesture detection sequence ongoing so that gestures
+        // are not detected by engine itself.
+        if (auto* gestureController = wpe_view_get_gesture_controller(m_view.wpeView()))
+            wpe_gesture_controller_cancel(gestureController);
```
You removed the `wpeView()` null check

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -397,6 +397,63 @@ Vector<WebKit::WebPlatformTouchPoint> ViewPlatform::touchPointsForEvent(WPEEvent
+    case WPE_GESTURE_NONE:
+        break;
+    case WPE_GESTURE_TAP:
+        if (double x, y; wpe_gesture_controller_get_gesture_position(gestureController, &x, &y)) {
```
So, you check it's not DOWN, but it can still be MOVE, we only want to handle tap on UP

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -397,6 +397,63 @@ Vector<WebKit::WebPlatformTouchPoint> ViewPlatform::touchPointsForEvent(WPEEvent
+            ));
+            auto phase = previousGesture != gesture
+                ? WebWheelEvent::Phase::PhaseBegan
+                : (wpe_gesture_controller_is_sequence_ended(gestureController)) ? WebWheelEvent::Phase::PhaseEnded : WebWheelEvent::Phase::PhaseChanged;
```
Can we handle the drag begin in the controller too? and add a way to query it, `wpe_gesture_controller_is_drag_begin` for example

carlosgcampos replied:
Do we really need the `is_sequence_ended`? I guess we could just check if the event is UP, no?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,144 @@
+ */
+void wpe_gesture_controller_handle_event(WPEGestureController* controller, WPEEvent* event)
+{
+    g_return_if_fail(controller);
```
Add another `g_return_if_fail` to check event is not nullptr either

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.h:
```diff
@@ -0,0 +1,87 @@
+    gboolean    (* get_gesture_delta)             (WPEGestureController *controller,
+                                                   double               *x,
+                                                   double               *y);
+    gboolean    (* is_sequence_ended)             (WPEGestureController *controller);
```
There are extra spaces between function name and parameters

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.h:
```diff
@@ -0,0 +1,87 @@
+WPE_API gboolean    wpe_gesture_controller_get_gesture_delta      (WPEGestureController *controller,
+                                                                   double               *x,
+                                                                   double               *y);
+WPE_API gboolean    wpe_gesture_controller_is_sequence_ended      (WPEGestureController *controller);
```
Ans here too

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,144 @@
+ */
+gboolean wpe_gesture_controller_get_gesture_position(WPEGestureController* controller, double* x, double* y)
+{
+    g_return_val_if_fail(controller, FALSE);
```
If out parameters are not optional, then we should add `g_return` to check they are not nullptr.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp:
```diff
@@ -0,0 +1,144 @@
+ */
+gboolean wpe_gesture_controller_get_gesture_delta(WPEGestureController* controller, double* x, double* y)
+{
+    g_return_val_if_fail(controller, FALSE);
```
Same here

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.cpp:
```diff
@@ -0,0 +1,85 @@
+
+void GestureDetector::handleEvent(WPEEvent* event)
+{
+    if (!event || (m_sequenceId && *m_sequenceId != wpe_event_touch_get_sequence_id(event)))
```
Can event be nullptr here?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.cpp:
```diff
@@ -0,0 +1,85 @@
+        break;
+    case WPE_EVENT_TOUCH_MOVE:
+        if (double x, y; wpe_event_get_position(event, &x, &y) && m_position) {
+            if (m_gesture != WPE_GESTURE_DRAG && std::hypot(x - m_position->x, y - m_position->y) > dragActivationThresholdPx) {
+                m_gesture = WPE_GESTURE_DRAG;
+                m_nextDeltaReferencePosition = m_position;
```
So, here is when we know it's a drag begin

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureController.h:
```diff
@@ -0,0 +1,87 @@
+    WPE_GESTURE_DRAG,
+} WPEGesture;
+
+G_BEGIN_DECLS
```
Move this right after the header includes

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.h:
```diff
@@ -0,0 +1,41 @@
+ */
+
+#ifndef WPEGestureControllerImpl_h
+#define WPEGestureControllerImpl_h
```
This is private, you can use `#pragma once`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.cpp:
```diff
@@ -0,0 +1,99 @@
+#include "config.h"
+#include "WPEGestureControllerImpl.h"
+
+#include <wpe/WPEGestureDetector.h>
```
```suggestion
#include "WPEGestureDetector.h"
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.h:
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <optional>
+#include <wpe/WPEEvent.h>
```
```suggestion
#include "WPEEvent.h"
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.h:
```diff
@@ -0,0 +1,62 @@
+
+class GestureDetector final {
+public:
+    void handleEvent(WPEEvent* event);
```
```suggestion
    void handleEvent(WPEEvent*);
```

Scony committed:
[WPE] WPE Platform: add support for tap and drag gestures
https://bugs.webkit.org/show_bug.cgi?id=265637

Reviewed by Carlos Garcia Campos.

This change extends the WPE platform API by introducing:
 - WPEGestureController - an interface (with default cross-WPE-platform implementation) that is used to process touch events in order to detect gestures.
This change also provides a code that:
 - Feeds WPEGestureController provided by WPE platform with events.
 - Handles events detected by WPEGestureController:
   - Tap gesture is handled as a mouse click.
   - Drag gesture is handled as a mouse wheel scroll.

* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/libwpe/WebEventFactory.h:
* Source/WebKit/Shared/wpe/NativeWebWheelEventWPE.cpp:
(WebKit::NativeWebWheelEvent::NativeWebWheelEvent):
* Source/WebKit/Shared/wpe/WebEventFactoryWPE.cpp:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::ViewPlatform):
(WKWPE::ViewPlatform::handleGesture):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp: Added.
(wpe_gesture_controller_default_init):
(wpe_gesture_controller_handle_event):
(wpe_gesture_controller_cancel):
(wpe_gesture_controller_get_gesture):
(wpe_gesture_controller_get_gesture_position):
(wpe_gesture_controller_get_gesture_delta):
(wpe_gesture_controller_is_drag_begin):
* Source/WebKit/WPEPlatform/wpe/WPEGestureController.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.cpp: Added.
(wpeHandleEvent):
(wpeCancel):
(wpeGetGesture):
(wpeGetGesturePosition):
(wpeGetGestureDelta):
(wpeIsDragBegin):
(wpe_gesture_controller_impl_class_init):
(wpe_gesture_controller_interface_init):
(wpeGestureControllerImplNew):
* Source/WebKit/WPEPlatform/wpe/WPEGestureControllerImpl.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.cpp: Added.
(WPE::GestureDetector::handleEvent):
(WPE::GestureDetector::reset):
* Source/WebKit/WPEPlatform/wpe/WPEGestureDetector.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_set_gesture_controller):
(wpe_view_get_gesture_controller):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: https://commits.webkit.org/280661@main
